### PR TITLE
Some test enhance, comments enhance and duplicate code reduce

### PR DESCRIPTION
**Pull request by shiywang**
while I was learning the codebase, found some place could be enhance, not confident about the comments of `Decode`  I changed isn't right.

/assign @sttts   PTAL, thanks

```release-note
NONE
```

### CODE REVIEW

**Comment by shiywang** [comment]
/assign @sttts 
could you please take a look ? thanks

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 1]
```diff
@@ -128,20 +131,24 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 		return nil, nil, err
 	}
 
-	if gvk != nil {
-		// apply kind and version defaulting from provided default
+	// apply kind and version defaulting from provided default
+	gvkDefault := func(actual *schema.GroupVersionKind, defaultGVK schema.GroupVersionKind) {
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: how about `completeGVK`?

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: in other words: gvkDefault should be an imperative with an obvious verb.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 2]
```diff
@@ -99,10 +99,13 @@ func init() {
...
-// extracted and no decoding will be performed. If into is not registered with the typer, then the object will be straight decoded using
-// normal JSON/YAML unmarshalling. If into is provided and the original data is not fully qualified with kind/version/group, the type of
-// the into will be used to alter the returned gvk. On success or most errors, the method will return the calculated schema kind.
+// load that data into an object matching the desired schema kind or the provided into.
+// If into is *runtime.Unknown, the raw data will be extracted and no decoding will be performed.
+// If into is not registered with the typer, then the object will be straight decoded using normal JSON/YAML unmarshalling.
+// If into is provided and the original data is not fully qualified with kind/version/group, the type of the into will be used to alter the returned gvk.
+// If into is nil or data's gvk different from into's gvk, it will generate a new Object with ObjectCreater.New(gvk)
+// On success or most errors, the method will return the calculated schema kind.
+// The gvk calculate priority will be originalData > default gvk > into
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: was the comment out of date or why was text dropped?

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go (hunk 1 above)
Comment (shiywang) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: done

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go (hunk 2 above)
Comment (shiywang) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: green line 102-105 && 107 is exactly same as previous except the indent format I changed each  `into` situation occupy one independent line.

106 && 108 is two new lines I added, just some complementary information of another `into` situation and schema kind calculated strategy, not sure those two lines is appropriate or not, if not I could delete those.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 3]
```diff
@@ -128,20 +131,24 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
...
 		}
-		if len(actual.Version) == 0 && actual.Group == gvk.Group {
-			actual.Version = gvk.Version
+		if len(actual.Version) == 0 && actual.Group == defaultGVK.Group {
+			actual.Version = defaultGVK.Version
 		}
 	}
 
+	if gvk != nil {
+		completeGVK(actual,*gvk)
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: gofmt

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 4]
```diff
@@ -128,20 +131,24 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 		return nil, nil, err
 	}
 
-	if gvk != nil {
-		// apply kind and version defaulting from provided default
+	// apply kind and version defaulting from provided default
+	completeGVK := func (actual *schema.GroupVersionKind, defaultGVK schema.GroupVersionKind) {
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: just noticed that this is just a `schema.GroupVersionKind`. So better write this functionally and move the helper func out:

```
func gvkWithDefaults(gvk schema.GroupVersionKind, defaults) schema.GroupVersionKind
```

**Comment by shiywang** [comment]
@sttts PTAL, thanks

**Pull request merged**